src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/UniversalCryptoOneShot.cs

```diff
@@ -20,31 +20,137 @@ internal static class UniversalCryptoOneShot
             }
 
+            // At this point the destination might be big enough but we don't know until we've
+            // transformed the last block, and input is within one block of being the right
+            // size.
+            // For sufficiently small ciphertexts, do them on the stack.
+            const int MaxInStackDecryptionBuffer = 128;
+            Span<byte> stackBuffer = stackalloc byte[MaxInStackDecryptionBuffer];
+
+            if (input.Length <= MaxInStackDecryptionBuffer)
+            {
+                int stackTransformFinal = cipher.TransformFinal(input, stackBuffer);
+                int depaddedLength = SymmetricPadding.GetPaddingLength(
+                    stackBuffer.Slice(0, stackTransformFinal),
+                    paddingMode,
+                    cipher.BlockSizeInBytes);
+
+                if (output.Length < depaddedLength)
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+
+                stackBuffer.Slice(0, depaddedLength).CopyTo(output);
+                bytesWritten = depaddedLength;
+                return true;
+            }
+
+            // If the source and destination do not overlap, we can decrypt directly in to the user buffer.
+            if (!input.Overlaps(output))
+            {
+                // At this point we know that we have multiple blocks that need to be decrypted.
+                // So we decrypt all but the last block directly in to the buffer. The final
+                // block we decrypt in to a stack buffer, and if it fits, copy the last block to
+                // the output.
+
+                // We should only get here if we have multiple blocks to transform. The single
+                // block case should have happened on the stack.
+                Debug.Assert(input.Length > cipher.BlockSizeInBytes);
+
+                int writtenToOutput = 0;
+
+                // CFB8 means this may not be an exact multiple of the block size.
+                // If the an AES CFB8 ciphertext length is 129 with PKCS7 padding, then
+                // we'll have 113 bytes in the unpaddedBlocks and 16 in the paddedBlock.
+                // We still need to do this on block size, not padding size. The CFB8
+                // padding byte might be block size. We don't want unpaddedBlocks to
+                // contain removable padding, so split on block size.
+                ReadOnlySpan<byte> unpaddedBlocks = input[..^cipher.BlockSizeInBytes];
+                ReadOnlySpan<byte> paddedBlock = input[^cipher.BlockSizeInBytes..];
+                Debug.Assert(paddedBlock.Length % cipher.BlockSizeInBytes == 0);
+                Debug.Assert(paddedBlock.Length <= MaxInStackDecryptionBuffer);
+
+                try
+                {
+                    writtenToOutput = cipher.Transform(unpaddedBlocks, output);
+                    int finalTransformWritten = cipher.TransformFinal(paddedBlock, stackBuffer);
```

Comment:
Do we need this split?  It feels like the holdback for padding is already doing this.

Unless we're below that layer, and one block in makes for one block out?

Reply:
It took me a minute to remember what the "holdback" you were referring to was, but I think I got it. You're referring to what's uni `UniversalCryptoDecryptor` and indeed we are below that.

https://github.com/dotnet/runtime/blob/95a68f229a11907358a6c860815c8af7610b86bf/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/UniversalCryptoDecryptor.cs#L191

We're operating on a lite cipher here, and the lite cipher is, well, lite. It doesn't really do anything besides hold some platform handle and have methods that call in to native APIs with that key handle. So for here, we need the split.